server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/Saml2BearerGrantAuthenticationConverter.java

```diff
@@ -0,0 +1,347 @@
+     * {@link Authentication} token into a {@link Saml2Authentication}
+     *
+     * @return the default response authentication converter strategy
+     */
+    static Converter<OpenSaml4AuthenticationProvider.AssertionToken, AbstractAuthenticationToken> createDefaultAssertionAuthenticationConverter() {
+        return assertionToken -> {
+            Assertion assertion = assertionToken.getAssertion();
+            Saml2AuthenticationToken token = assertionToken.getToken();
+            String username = assertion.getSubject().getNameID().getValue();
+            Map<String, List<Object>> attributes = OpenSaml4AuthenticationProvider.getAssertionAttributes(assertion);
+            List<String> sessionIndexes = OpenSaml4AuthenticationProvider.getSessionIndexes(assertion);
+            DefaultSaml2AuthenticatedPrincipal principal = new DefaultSaml2AuthenticatedPrincipal(username, attributes,
+                    sessionIndexes);
+            String registrationId = token.getRelyingPartyRegistration().getRegistrationId();
+            principal.setRelyingPartyRegistrationId(registrationId);
+            return new Saml2Authentication(principal, token.getSaml2Response(),
+                    AuthorityUtils.createAuthorityList("ROLE_USER"));
+        };
+    }
+
+    /**
+     * Construct a default strategy for validating each SAML 2.0 Assertion and associated
+     * {@link Authentication} token
+     *
+     * @param validationContextParameters a consumer for editing the values passed to the
+     *                                    {@link ValidationContext} for each assertion being validated
+     * @return the default assertion validator strategy
+     * @since 5.8
+     */
+
+    @Override
+    public Authentication convert(HttpServletRequest request) throws AuthenticationException {
+        RelyingPartyRegistration relyingPartyRegistration = relyingPartyRegistrationResolver.resolve(request, null);
+
+        String serializedAssertion = request.getParameter("assertion");
+        byte[] decodedAssertion = Saml2Utils.samlDecode(serializedAssertion);
+        String assertionXml = new String(decodedAssertion, StandardCharsets.UTF_8);
+
+        Assertion assertion = parseAssertion(assertionXml);
+        Saml2AuthenticationToken authenticationToken = new Saml2AuthenticationToken(relyingPartyRegistration, assertionXml);
+        process(authenticationToken, assertion);
+
+        String subjectName = assertion.getSubject().getNameID().getValue();
+        String alias = relyingPartyRegistration.getRegistrationId();
+        IdentityZone zone = identityZoneManager.getCurrentIdentityZone();
+
+        UaaPrincipal initialPrincipal = new UaaPrincipal(NotANumber, subjectName, subjectName,
+                alias, subjectName, zone.getId());
+
+        boolean addNew;
+        IdentityProvider<SamlIdentityProviderDefinition> idp;
+        SamlIdentityProviderDefinition samlConfig;
+        try {
```

Comment:
a 2nd lookup in DB. right now dont have a better idea, but I would do it earlier, before signature validation